File: Mac/PictureSorter/ViewController.m
//
//  ViewController.m
//  PictureSorter
//
//  Created by Zenel Kazushi on 8/18/16.
//  Copyright (c) 2016 Zenel Kazushi. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@property (weak) IBOutlet NSImageView *imageViewOutput;
@property (nonatomic, strong) NSString *inputPath;
@property (nonatomic, strong) NSArray *imageFiles;
@property (nonatomic, assign) int index;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    // Do any additional setup after loading the view.
}

- (void)setRepresentedObject:(id)representedObject {
    [super setRepresentedObject:representedObject];

    // Update the view, if already loaded.
}

- (void)keyDown:(NSEvent *)theEvent {
    if (theEvent.type != NSKeyDown || !self.imageFiles) {
        return;
    }
    NSString *folderName = nil;
    switch ([theEvent.characters characterAtIndex:0]) {
        case '1':
            folderName = @"1";
            break;
        case '2':
            folderName = @"2";
            break;
        case '3':
            folderName = @"3";
            break;
        case '4':
            folderName = @"4";
            break;
        default:
            return;
    }
    [self copyTopFileTo:folderName];
}

- (void)copyTopFileTo:(NSString *)folderName {
    NSString *fullFolderPath = [self.inputPath stringByAppendingPathComponent:folderName];
    NSString *fileName = [self.imageFiles[self.index] lastPathComponent];
    [[NSFileManager defaultManager] moveItemAtPath:self.imageFiles[self.index] toPath:[fullFolderPath stringByAppendingPathComponent:fileName] error:nil];
    ++self.index;
    [self showImage];
}

- (IBAction)buttonSelectFolder:(id)sender {
    NSOpenPanel *openPanel = [NSOpenPanel openPanel];
    [openPanel setCanChooseDirectories:YES];
    [openPanel setCanChooseFiles:YES];
    [openPanel setCanCreateDirectories:NO];
    [openPanel setAllowsMultipleSelection:NO];
    NSInteger clicked = [openPanel runModal];
    
    if (clicked == NSFileHandlingPanelOKButton) {
        self.inputPath = [[openPanel URL] path];
        [self startShowingPictures];
    }
}

- (void)startShowingPictures {
    NSArray *files = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:self.inputPath error:nil];
    NSMutableArray *imageFiles = [NSMutableArray new];
    for (NSString *file in files) {
        if ([[file pathExtension] isEqualToString:@"png"]
            || [[file pathExtension] isEqualToString:@"jpg"]
            || [[file pathExtension] isEqualToString:@"gif"]) {
            [imageFiles addObject:[self.inputPath stringByAppendingPathComponent:file]];
        }
    }
    self.imageFiles = [imageFiles copy];
    [[NSFileManager defaultManager] createDirectoryAtPath:[self.inputPath stringByAppendingPathComponent:@"1"] withIntermediateDirectories:NO attributes:nil error:nil];
    [[NSFileManager defaultManager] createDirectoryAtPath:[self.inputPath stringByAppendingPathComponent:@"2"] withIntermediateDirectories:NO attributes:nil error:nil];
    [[NSFileManager defaultManager] createDirectoryAtPath:[self.inputPath stringByAppendingPathComponent:@"3"] withIntermediateDirectories:NO attributes:nil error:nil];
    [[NSFileManager defaultManager] createDirectoryAtPath:[self.inputPath stringByAppendingPathComponent:@"4"] withIntermediateDirectories:NO attributes:nil error:nil];
    self.index = 0;
    [self showImage];
}

- (void)showImage {
    if (self.index >= self.imageFiles.count) {
        NSLog(@"All done!");
        self.imageViewOutput.image = nil;
        return;
    }
    NSString *file = self.imageFiles[self.index];
    NSImage *image = [[NSImage alloc] initWithContentsOfFile:file];
    self.imageViewOutput.image = image;
}

@end

File: Mac/PictureSorter/ViewController.swift

//
//  ViewController.swift
//  PictureSorter
//
//  Created by Zenel Kazushi on 9/9/17.
//  Copyright © 2017 Zenel Kazushi. All rights reserved.
//

import Cocoa

class ViewController: NSViewController {
    @IBOutlet var imageViewPreview: NSImageView!
    @IBOutlet var buttonSelectFolder: NSButton!
    var inputPath: URL?
    var imageFiles: [URL]?
    var currentImageIndex: Int = -1
    
    override func keyDown(with event: NSEvent) {
        if event.type != .keyDown || imageFiles == nil {
            return
        }
        guard let name = folderName(for: event) else {
            return
        }
        copyFile(to: name)
    }
    
    private static func isImage(_ path: URL) -> Bool {
        guard let uti = UTTypeCreatePreferredIdentifierForTag(kUTTagClassFilenameExtension, path.pathExtension as CFString, nil)?.autorelease() else {
            fatalError("Could not build UTI Type!")
        }
        return UTTypeConformsTo(uti.takeUnretainedValue(), kUTTypeImage)
    }
    
    @IBAction func onButtonSelectFolderClicked(_ sender: NSButton) {
        let openPanel = NSOpenPanel()
        openPanel.canChooseDirectories = true
        openPanel.canChooseFiles = true
        openPanel.canCreateDirectories = false
        openPanel.allowsMultipleSelection = false
        let clicked = openPanel.runModal()
        if clicked == NSFileHandlingPanelOKButton {
            inputPath = openPanel.url
            guard inputPath != nil else {
                fatalError("Open panel succeeded, but selected no paths!")
            }
            startShowingPictures()
        }
    }
    
    private func startShowingPictures() {
        guard let inputPath = inputPath else {
            fatalError("Logic error: tried to start showing pictures without an input path!")
        }
        let files = try! FileManager.default.contentsOfDirectory(at: inputPath, includingPropertiesForKeys: [URLResourceKey]())
        var imageFiles = [URL]()
        for file in files {
            if ViewController.isImage(file) {
                imageFiles.append(file)
            }
        }
        self.imageFiles = imageFiles
        currentImageIndex = 0
        showImage()
    }
    
    private func folderName(for event: NSEvent) -> String? {
        guard let chars = event.characters?.characters, chars.count == 1 else {
            return nil
        }
        let validNames = [ "1", "2", "3", "4", "5", "6", "7", "8", "9", "0" ]
        let eventChar = String(chars[chars.index(chars.startIndex, offsetBy: 1)])
        guard validNames.contains(eventChar) else {
            return nil
        }
        return eventChar
    }
    
    private func copyFile(to folderName: String) {
        guard let inputPath = inputPath else {
            fatalError("Logic error: copyFile(to:) called with no input path set!")
        }
        let fullFolderPath = inputPath.appendingPathComponent(folderName, isDirectory: true)
        guard let imageFiles = imageFiles else {
            fatalError("Logic error: image files are not accessible!")
        }
        let fileURL = imageFiles[currentImageIndex]
        let fileName = fileURL.lastPathComponent
        let newURL = fullFolderPath.appendingPathComponent(fileName)
        
        try! FileManager.default.moveItem(at: fileURL, to: newURL)
        currentImageIndex += 1
        showImage()
    }
    
    private func showImage() {
        guard let imageFiles = imageFiles else {
            fatalError("Logic error: imageFiles is nil!")
        }
        if currentImageIndex >= imageFiles.count {
            print("All done!")
            imageViewPreview.image = nil
            return
        }
        imageViewPreview.image = NSImage(contentsOf: imageFiles[currentImageIndex])
    }
}

